### PR TITLE
Fix Host Keepalive serial output with multi-serial

### LINKED ISSUE #21010: [BUG] Random freezes during printing over serial connection (skr 1.4 turbo)
### Bug Description

I'm experiencing radom freezes when printing with Octoprint over serial connection.
Using SKR 1.4 turbo, RPI4, directly connected to TFT serial interface (SERIAL_PORT 0).
Running bugfix-2x 004bed8a7fc3ff9feb73a0ea9794635b50073c27
I am not able to identify any specific circumstances of the error - sometimes it happens on first layer, and sometimes after few hours of printing the same gcode.
Print head stops, screen (ulti controler) stops refreshing, communication freezes (there is "timeout" message in octoprint terminal window).

I'm suspecting it's related to recent serial changes, because:
- I'm not experiencing this issue when printing from SD card
- I've reverted to 811bb7997cf34f6d9559c5f281d209445c4fc123 and so far it seems to be working fine.

### Configuration Files

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5936904/Configuration.zip)

### PR DESCRIPTION
**Pull request by rhapsodyv**
### Description

This is a continuation of the stress tests I'm doing on multi serial, to fix corner cases.

I found an issue that happens in a very specific combination of events, but that can happen very often on a real printing operation: multi serial + idle stacking (slow commands that call idle until finished) + keep alive code.

When marlin is executing some slow commands, it may stay in a loop waiting for it to complete. Inside that loop, marlin may call idle periodically. So, the idle will check for new serial commands (in the other serial) to enqueue. When it receive the command, it will reply "busy" (on keep alive function), but it send to the wrong serial port.

Call stack:
```
idle
  process_next_command
    Gxxx
       while(something)
          idle
            keepalive (sent to the current serial)
            other serial tries to send data, thinks it succeeds, but marlin simply doesn't reply (until the actual commands)
            data loss and wrong behavior on the first serial (it will cause resend on serial 1)
```

I could simulate resend commands on octoprint this way. And I fixed making the keep alive reply "busy" for all serial ports... 

That in fact it seems correct, because marlin will only handle one command at time, so the keep alive should warns all serial ports that marlin is busy, to avoid timeout or false-resend commands everywhere.

### Benefits

More robust multi serial.
May fix #21010 and #21244

### Related Issues

#21249

### CODE REVIEW

**Comment by thinkyhead** [comment]
It looks like `multiSerial.portMask` is being left uninitialized (zero) so the fallback for `SERIAL_REDIRECT` is no ports. The fallback should probably be `0xFF`, or `SERIAL_ALL`, and this might actually be the correct fix.

**Comment by rhapsodyv** [comment]
> It looks like `multiSerial.portMask` is being left uninitialized (zero) so the fallback for `SERIAL_REDIRECT` is no ports. The fallback should probably be `0xFF`, or `SERIAL_ALL`, and this might actually be the correct fix.

yes, can be. But remember that the second idle is called _inside_ another running command.... so, it may set the serial port to the current command... and the other command just hang (it's what happens).

**Comment by rhapsodyv** [comment]
```log
Send: N4848 G1 X2.372 Y2.8189999*30
Recv: ok
Send: N4849 G1 X2.372 Y2.8189999*31
Recv:  T:-15.00 /0.00 @:0
Recv: echo:busy: processing
Printer seems to support the busy protocol, will adjust timeouts and set busy interval accordingly
Recv:  T:-15.00 /0.00 @:0
Recv: ok
Send: N4850 M113 S2*88
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-14.92 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Communication timeout while printing, trying to trigger response from printer. Configure long running commands or increase communication timeout if that happens regularly on specific commands or long moves.
Send: N4851 M105*31
Recv: Error:checksum mismatch, Last Line: 4849
Recv: Resend: 4850
Send: N4850 M113 S2*88
Recv: ok
Send: N4851 M105*31
Recv: ok T:-15.00 /0.00 @:0
Send: N4852 G1 X2.372 Y2.8189999*21
```

**Comment by rhapsodyv** [comment]
Test:

Octo "printing" in one serial 3 (only g1 x2 x2)
Screen Shot 2021-03-08 at 01 18 34

Then, in another serial (serial 1), I send a long move, to generate a wait on planner, to it call idle and get a keep alive:
Screen Shot 2021-03-08 at 01 18 48


This is what happens on octo:
```log
Send: N4848 G1 X2.372 Y2.8189999*30
Recv: ok
Send: N4849 G1 X2.372 Y2.8189999*31
Recv:  T:-15.00 /0.00 @:0
Recv: echo:busy: processing. <<<<-receive the busy
Printer seems to support the busy protocol, will adjust timeouts and set busy interval accordingly
Recv:  T:-15.00 /0.00 @:0
Recv: ok
Send: N4850 M113 S2*88
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-14.92 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0
Recv:  T:-15.00 /0.00 @:0

Communication timeout while printing, trying to trigger response from printer. Configure long running commands or increase communication timeout if that happens regularly on specific commands or long moves.

Send: N4851 M105*31
Recv: Error:checksum mismatch, Last Line: 4849
Recv: Resend: 4850
Send: N4850 M113 S2*88
Recv: ok
Send: N4851 M105*31
Recv: ok T:-15.00 /0.00 @:0

Send: N4852 G1 X2.372 Y2.8189999*21
```

After adding the like that send `busy` for all serial ports, octo print will get hang or get timeout

**Comment by rhapsodyv** [comment]
This behaviour happened always in my tests: every time I generate a wait having commands for another serial, one serial loses a command and get timeout. With busy broadcast, I got no more `resend` erros on octo, even if I send commands in loop to the two serials, generating lots of busy, etc.

**Comment by thinkyhead** [comment]
But, `PORT_REDIRECT` carries its own "stack" with it to ensure that everything is using the port it expects. It creates an object that restores the previous serial mask when it goes out of scope. If it isn't working now, maybe it's not restoring the previous value properly.

**Comment by rhapsodyv** [comment]
> But, `SERIAL_REDIRECT` carries its own "stack" with it to ensure that everything is using the port it expects. It creates an object that restores the previous serial mask when it goes out of scope. If it isn't working now, maybe it's not restoring the previous value properly.

Yes, but its value will be the last value set, right? That is the value of the current command being executed. 

You can easy test it:
in the serial 1, send a M0. This serial will be receiving `paused by the user` every "keep alive seconds"

then, In the serial 2, send a M114... Marlin will not respond anything to this serial, never..... even if, in fact, marlin is currently busy... So, this serial will timeout, if the command is queued... so, after the serial 1 release, marlin will execute the M114, and the host may send it again!!

That is the reason I sent the PR, because makes sense, for me, that keep alive should broadcast (as all host commands).

But yes, maybe we have more hidden issues hehehe...

**Comment by thinkyhead** [comment]
The current command does redirect its output, and presumably it wants to also redirect the output of anything within `idle` that is intended as a message about the current state. Things called by `idle()` will be responsible for redirecting to `SERIAL_ALL` if they need to. When the running command is complete it should be restoring the serial port mask back to the default (i.e., `SERIAL_ALL`).

**Comment by thinkyhead** [comment]
…so you are correct to redirect the output of keepalive.

**Comment by thinkyhead** [comment]
And now, to fix this exp32 compile before I go ahead and merge….

**Comment by rhapsodyv** [comment]
I found ... the lib `arduinoWebSockets` was updated to use the lasts code from `espressif32`... but we are using `espressif32@2.1.0`.... Or we try to use old version of `arduinoWebSockets` or new version of  `espressif32`...

we can point `arduinoWebSockets` to one version old and it will work... 
On features, point arduinoWebSockets to 2.3.4 (current is 2.3.5, committed yesterday)
```
arduinoWebSockets=https://github.com/Links2004/arduinoWebSockets/archive/2.3.4.zip
```

File: Marlin/src/core/serial.h [hunk 1]
```diff
@@ -62,7 +62,7 @@ extern uint8_t marlin_debug_flags;
 //
 // Serial redirection
 //
-#define SERIAL_ALL 0x7F
+#define SERIAL_ALL 0xFF
```
Comment (rhapsodyv) [inline Marlin/src/core/serial.h:65]: Hum... The default as `0x7F` don't seem to be a mistake... it appears to be intentional ... @X-Ryl669?

Comment (thinkyhead) [reply Marlin/src/core/serial.h:65]: It was only renamed in his serial refactor. It used to represent a non-negative but unlikely serial INDEX (plus a BOTH value). Now it represents a MASK of up to 8 serial ports. It could also just be defined as `(1 << NUM_SERIAL) - 1`.

Comment (X-Ryl669) [reply Marlin/src/core/serial.h:65]: That's because the serial index is signed and negative values have a special meaning (don't remember why right now, but they are few places where -1 is used for the serial index default value). So it's set to  7 bits (= 7 serial ports) and not 8 bits here.

File: Marlin/src/core/serial_hook.h [hunk 2]
```diff
@@ -165,12 +165,12 @@ struct RuntimeSerial : public SerialBase< RuntimeSerial<SerialT> >, public Seria
...
 };
 
-// A class that's duplicating its output conditionally to 2 serial interface
+// A class that duplicates its output conditionally to 2 serial interfaces
 template <class Serial0T, class Serial1T, const uint8_t offset = 0, const uint8_t step = 1>
 struct MultiSerial : public SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > {
   typedef SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > BaseClassT;
 
-  uint8_t    portMask;
+  uint8_t    portMask = 0xFF;
```
Comment (X-Ryl669) [inline Marlin/src/core/serial_hook.h]: You don't need this, it's set in the constructor to the actual required mask (don't use full mask here, since you can chain Multiserial, they **must** have a different portMask)

File: Marlin/src/core/serial.h (hunk 1 above)
Comment (X-Ryl669) [reply Marlin/src/core/serial.h:65]: If you want to set it to 0xFF, then you'll need to change to `typedef uint8_t serial_index_t`, IMHO.

Comment (thinkyhead) [reply Marlin/src/core/serial.h:65]: `SERIAL_ALL` is not a serial index. It has always been used like a serial mask (previously could be considered as having 2 bits) and in examining the code I see no instance where is it used in the context of a serial index. So I think we're all good.

File: Marlin/src/core/serial_hook.h (hunk 2 above)
Comment (thinkyhead) [reply Marlin/src/core/serial_hook.h]: Ah, right. I see the initializer now. Thanks!

File: Marlin/src/core/serial.h (hunk 1 above)
Comment (X-Ryl669) [reply Marlin/src/core/serial.h:65]: Yep, but it's saved into a serial_index_t in the `PORT_REDIRECT` macro. And since the value `-1` is used as a default argument (in queue.cpp), setting `SERIAL_ALL` to 255 (which is -1), means that default value for the queue's functions now means all serials.

I'm not fond of mixing index and mask here. Either we should move to mask everywhere, or remove ambiguities (the worst that could happen is doing a `BV(serial_index)` when `serial_index` is -1). It'll be very hard to track in a debugger where does the -1 comes from if it can come from default parameter **and** `SERIAL_ALL`.

Comment (X-Ryl669) [reply Marlin/src/core/serial.h:65]: See queue.h line 82 and 88. I think it's mixing index and mask here. It should read `serial_index_t p = 0` instead of `-1` since that would redirect to the first serial port when not in multiserial case. And when this is done, serial_index_t could be changed to an unsigned value (makes more sense to me, IMHO).

**Comment by adolson** [comment]
bugfix-2.0.x fails to compile in Arduino IDE after this merge. Was working yesterday. How do I fix this?

```
In file included from sketch/src/HAL/ESP32/FlushableHardwareSerial.cpp:23:0:
sketch/src/HAL/ESP32/FlushableHardwareSerial.h: In constructor ‘FlushableHardwareSerial::FlushableHardwareSerial(int)’:
sketch/src/HAL/ESP32/FlushableHardwareSerial.h:31:64: error: no matching function for call to ‘HardwareSerial::HardwareSerial(int&)’
   FlushableHardwareSerial(int uart_nr) : HardwareSerial(uart_nr) {}
                                                                ^
In file included from sketch/src/HAL/ESP32/FlushableHardwareSerial.h:24:0,
                 from sketch/src/HAL/ESP32/FlushableHardwareSerial.cpp:23:
/usr/share/arduino/hardware/arduino/avr/cores/arduino/HardwareSerial.h:117:12: note: candidate: HardwareSerial::HardwareSerial(volatile uint8_t*, volatile uint8_t*, volatile uint8_t*, volatile uint8_t*, volatile uint8_t*, volatile uint8_t*)
     inline HardwareSerial(
            ^
/usr/share/arduino/hardware/arduino/avr/cores/arduino/HardwareSerial.h:117:12: note:   candidate expects 6 arguments, 1 provided
/usr/share/arduino/hardware/arduino/avr/cores/arduino/HardwareSerial.h:93:7: note: candidate: constexpr HardwareSerial::HardwareSerial(const HardwareSerial&)
 class HardwareSerial : public Stream
       ^
/usr/share/arduino/hardware/arduino/avr/cores/arduino/HardwareSerial.h:93:7: note:   no known conversion for argument 1 from ‘int’ to ‘const HardwareSerial&’
/usr/share/arduino/hardware/arduino/avr/cores/arduino/HardwareSerial.h:93:7: note: candidate: constexpr HardwareSerial::HardwareSerial(HardwareSerial&&)
/usr/share/arduino/hardware/arduino/avr/cores/arduino/HardwareSerial.h:93:7: note:   no known conversion for argument 1 from ‘int’ to ‘HardwareSerial&&’
exit status 1
Error compiling for board Arduino Mega or Mega 2560.
```

**Comment by rhapsodyv** [comment]
@adolson take a look if it solves: https://github.com/rhapsodyv/Marlin/tree/21283-followup

**Comment by adolson** [comment]
Yes, that compiles successfully.
